File: 2019-2020_programmers/_10_/src/java/PalindromeRecursiveLoop.java
class PalindromeRecursiveLoop {
  public static void main(String[] args) {
    String[] data = new String[] {
      "abcdcba",
      "abacba", 
      "abacde"
    };
    System.out.println(solution(data[2]));
  }

  static int solution(String s) {
    return maxPalindrome(s, s.length());
  }
  static int maxPalindrome(String str, int len) {
    while (len > 1) {
      int midpoint = len / 2 - 1;
      for (int shift = 0; shift <= str.length() - len; shift++) {
        boolean isPalindrome = true;
        for (int cursor = 0; cursor <= midpoint; cursor++) {
          if (str.charAt(cursor + shift) != str.charAt(len - 1 - cursor + shift)) {
            isPalindrome = false;
            break;
          }
        }
        if (isPalindrome) return len;
      }
      len--;
    }
    return 1;
  }
}

File: 2019-2020_programmers/_10_/src/typescript/palindrome_recursive_loop.ts
let data2: String[] = ["abcdcba", "abacba", "abacde"];

const solution2 = s => {
  const maxPalindrome = (str, len) => {
    while (len > 1) {
      let midpoint = Math.ceil(len / 2) - 1;
      // console.log("len: " + len, "midpoint: " + midpoint);
      for (let shift = 0; shift <= str.length - len; shift++) {
        // console.log("string: " + str.substring(0 + shift, len + shift));
        let isPalindrome = true;
        for (let cursor = 0; cursor <= midpoint; cursor++) {
          // console.log(
          //   str[cursor + shift],
          //   str[cursor + shift] == str[len - 1 - cursor + shift] ? "==" : "!=",
          //   str[len - 1 - cursor + shift] + "\n"
          // );
          if (str[cursor + shift] != str[len - 1 - cursor + shift]) {
            isPalindrome = false;
            break;
          }
        }
        if (isPalindrome) return len;
      }
      len--;
    }
    return 1;
  };
  return maxPalindrome(s, s.length);
};

console.log(solution2(data2[2]));
